File: dart/comsrv/lib/gctcan.dart
import 'dart:typed_data';

import 'package:broadcast_wsrpc/lib.dart';
import 'package:comsrv/can.dart';

const msgTypeSysCtrl = 1;
const msgTypeMonitoringData = 7;
const msgTypeMonitoringRequest = 8;
const msgTypeDdp = 12;
const msgTypeHeartBeat = 14;
const maxDdpDataLenV1 = 61; // 8 message * 8bytes - crc - cmd
const maxDdpDataLenV2 = 8 * 256 - 3; // 256 message * 8bytes - crc - cmd

const broadcastAddress = 0x7F;

class MonitorValue {
  final int readingIndex;
  final int groupIndex;
  final Uint8List data;

  MonitorValue(this.readingIndex, this.groupIndex, this.data);
}

class GctCanDevice {
  int controllerId;
  final CanBus canBus;
  Duration timeout = Duration(milliseconds: 100);

  GctCanDevice(this.canBus, this.controllerId);

  Future<List<MonitorValue>> _receiveMonitorRequests(int destinationNodeId,
      int groupIndex, Uint8List readings, Stream<GctMessage> messages) async {
    final receivedReadings = <int, MonitorValue?>{};
    for (final reading in readings) {
      receivedReadings[reading] = null;
    }
    await for (final msg in messages) {
      if (msg is! MonitoringData ||
          msg.groupIndex != groupIndex ||
          msg.source != destinationNodeId ||
          !readings.contains(msg.readingIndex)) {
        continue;
      }
      final value = MonitorValue(msg.readingIndex, msg.groupIndex, msg.data);
      receivedReadings[msg.readingIndex] = value;
      if (!receivedReadings.containsValue(null)) {
        break;
      }
    }
    final ret = <MonitorValue>[];
    for (final reading in readings) {
      ret.add(receivedReadings[reading]!);
    }
    return ret;
  }

  Future<List<MonitorValue>> monitorRequest(
      int destinationNodeId, int groupIndex, Uint8List readings) async {
    final messages = canBus.gctMessages();
    canBus.sendGct(MonitoringRequest(
        source: controllerId,
        destination: destinationNodeId,
        groupIndex: groupIndex,
        readings: readings));
    return await _receiveMonitorRequests(
            destinationNodeId, groupIndex, readings, messages)
        .timeout(timeout);
  }

  Future<void> sysctrlWrite(
      int destinationNodeId, int command, Uint8List data) async {
    final txMsg = SysCtrlMessage(
        controllerId, destinationNodeId, command, SysCtrlType.none, data);
    await canBus.sendGct(txMsg);
  }

  Future<Uint8List> sysctrlRead(int destinationNodeId, int command) async {
    return await sysctrlWriteRead(destinationNodeId, command, Uint8List(0));
  }

  Future<Uint8List> sysctrlWriteRead(
      int destinationNodeId, int command, Uint8List data) async {
    final sysctrlMessages = canBus
        .gctMessages()
        .where((event) => event is SysCtrlMessage)
        .map((event) => event as SysCtrlMessage)
        .where((event) =>
            event.source == destinationNodeId &&
            event.command == command &&
            event.sysCtrlType == SysCtrlType.value);
    final txMsg = SysCtrlMessage(
        controllerId, destinationNodeId, command, SysCtrlType.query, data);
    await canBus.sendGct(txMsg);
    final retMsg = await sysctrlMessages.first.timeout(timeout);
    return retMsg.data;
  }

  Future<Uint8List> ddp(int destinationNodeId, Uint8List data,
      {int version = 2}) async {
    final ddpMessages = canBus
        .gctMessages()
        .where((event) => event is DdpMessage)
        .map((event) => event as DdpMessage)
        .where((event) =>
            event.source == destinationNodeId &&
            event.destination == controllerId &&
            event.version == version);
    await canBus
        .sendGct(DdpMessage(controllerId, destinationNodeId, data, version));
    final retMsg = await ddpMessages.first.timeout(timeout);
    return retMsg.data;
  }
}

abstract class GctMessage {
  int get source;
  int get destination;
  int get messageType;

  JsonObject toJson();

  static GctMessage? fromJson(JsonObject object) {
    if (object.containsKey("SysCtrl")) {
      return SysCtrlMessage.fromJson(object["SysCtrl"]);
    } else if (object.containsKey("MonitoringData")) {
      return MonitoringData.fromJson(object["MonitoringData"]);
    } else if (object.containsKey("MonitoringRequest")) {
      return MonitoringRequest.fromJson(object["MonitoringRequest"]);
    } else if (object.containsKey("Ddp")) {
      return DdpMessage.fromJson(object["Ddp"]);
    } else if (object.containsKey("Heartbeat")) {
      return Heartbeat.fromJson(object["Heartbeat"]);
    }
    return null;
  }
}

enum SysCtrlType { none, value, query }

class SysCtrlMessage extends GctMessage {
  final SysCtrlType sysCtrlType;
  final int _source;
  final int _destination;
  final int command;
  final Uint8List data;

  SysCtrlMessage(this._source, this._destination, this.command,
      this.sysCtrlType, this.data);

  factory SysCtrlMessage.fromJson(JsonObject msg) {
    final source = msg["src"];
    final destination = msg["dst"];
    final command = msg["cmd"];

    final tp = msg["tp"] as String;
    SysCtrlType? sysCtrlType;
    switch (tp) {
      case "None":
        sysCtrlType = SysCtrlType.none;
        break;
      case "Value":
        sysCtrlType = SysCtrlType.value;
        break;
      case "Query":
        sysCtrlType = SysCtrlType.query;
        break;
      default:
        throw ArgumentError("No such SysCtrl type");
    }
    final data = Uint8List.fromList(msg["data"].cast<int>());
    return SysCtrlMessage(source, destination, command, sysCtrlType, data);
  }

  @override
  int get destination => _destination;

  @override
  int get messageType => msgTypeSysCtrl;

  @override
  int get source => _source;

  @override
  JsonObject toJson() {
    String? tp;
    switch (sysCtrlType) {
      case SysCtrlType.none:
        tp = "None";
        break;
      case SysCtrlType.value:
        tp = "Value";
        break;
      case SysCtrlType.query:
        tp = "Query";
        break;
    }
    return {
      "SysCtrl": {
        "src": _source,
        "dst": _destination,
        "cmd": command,
        "tp": tp,
        "data": data.toList(),
      }
    };
  }
}

class MonitoringData extends GctMessage {
  final int _source;
  final int groupIndex;
  final int readingIndex;
  final Uint8List data;

  MonitoringData(this._source, this.groupIndex, this.readingIndex, this.data);

  factory MonitoringData.fromJson(JsonObject object) {
    final source = object["src"];
    final groupIndex = object["group_idx"];
    final readingIndex = object["readings_idx"];
    final data = Uint8List.fromList(object["data"]);
    return MonitoringData(source, groupIndex, readingIndex, data);
  }

  @override
  int get destination => broadcastAddress;

  @override
  int get messageType => msgTypeMonitoringData;

  @override
  int get source => _source;

  @override
  JsonObject toJson() {
    return {
      "MonitoringData": {
        "src": source,
        "group_idx": groupIndex,
        "reading_idx": readingIndex,
        "data": data.toList(),
      }
    };
  }
}

class MonitoringRequest extends GctMessage {
  int groupIndex;

  final int _source;
  final int _destination;
  final Uint8List readings;

  factory MonitoringRequest.fromJson(JsonObject object) {
    final int destination = object["dst"];
    final int source = object["source"];
    final readings = Uint8List.fromList(object["readings"]);
    final groupIndex = object["group_idx"];
    return MonitoringRequest(
      source: source,
      destination: destination,
      groupIndex: groupIndex,
      readings: readings,
    );
  }

  MonitoringRequest({
    required source,
    required destination,
    required this.groupIndex,
    required this.readings,
  })  : _source = source,
        _destination = destination;

  @override
  int get destination => _destination;

  @override
  int get source => _source;

  @override
  int get messageType => msgTypeMonitoringRequest;

  @override
  JsonObject toJson() {
    return {
      "MonitoringRequest": {
        "src": _source,
        "dst": _destination,
        "group_idx": groupIndex,
        "readings": readings,
      }
    };
  }
}

class Heartbeat extends GctMessage {
  final int _source;
  final int productId;

  Heartbeat(this._source, this.productId);

  factory Heartbeat.fromJson(JsonObject object) {
    return Heartbeat(object["src"], object["product_id"]);
  }

  @override
  int get destination => broadcastAddress;

  @override
  int get messageType => msgTypeHeartBeat;

  @override
  int get source => _source;

  @override
  JsonObject toJson() {
    return {
      "Heartbeat": {
        "src": _source,
        "product_id": productId,
      }
    };
  }
}

class DdpMessage extends GctMessage {
  final int _destination;
  final int _source;
  Uint8List data;
  int version;

  DdpMessage(this._source, this._destination, this.data, this.version);

  factory DdpMessage.fromJson(JsonObject object) {
    final int source = object["src"];
    final int destination = object["dst"];
    final int version = object["version"];
    final List<int> data = object["data"];

    return DdpMessage(source, destination, Uint8List.fromList(data), version);
  }

  @override
  int get destination => _destination;

  @override
  int get messageType => msgTypeDdp;

  @override
  int get source => _source;

  @override
  JsonObject toJson() {
    return {
      "Ddp": {
        "version": version,
        "src": _source,
        "dst": _destination,
        "data": data.toList(),
      }
    };
  }
}

File: python/comsrv/gctcan.py
import asyncio
import struct
from typing import Any, List, Set, Union, Iterable, Optional
from broadcast_wsrpc.client import Receiver  # type: ignore

from .can import (
    DdpMessage,
    SysCtrlMessage,
    SysCtrlType,
    MonitoringRequest,
    GctMessage,
    MonitoringData,
    CanBus,
)


class DdpFilter(object):
    def __init__(self, src: int, dst: int):
        self._src = src
        self._dst = dst

    def __call__(self, msg: GctMessage) -> bytes | None:
        if not isinstance(msg, DdpMessage):
            return None
        ok = msg.src == self._src and msg.dst == self._dst
        if not ok:
            return None
        return msg.data


CONTROLLER_NODEID = 0x7E
BROADCAST_ADDR = 0x7F
DEFAULT_DDP_TIMEOUT = 0.1


class GctCanBus(object):
    def __init__(self, bus: CanBus, ddp_version: int = 2) -> None:
        self._bus = bus
        self._version = ddp_version

    @property
    def can_bus(self) -> CanBus:
        return self._bus

    async def connect(self) -> "GctCanBus":
        await self._bus.connect()
        return self

    async def sysctrl_request(
        self,
        src: int,
        cmd: int,
        data: bytes | None = None,
        dst: int = BROADCAST_ADDR,
    ) -> bytes:
        """
        Read back sysctrl, optionally writing data to the endpoint.

        :param cmd: The command to send the data to
        :param data: The data to be sent to the command. If None, no data is sent and the data is only read back.
        :return: The data currently set at the specific command.
        """

        def flt(msg: GctMessage) -> SysCtrlMessage | None:
            if not isinstance(msg, SysCtrlMessage):
                return None
            if msg.src != dst and dst != BROADCAST_ADDR:
                return None
            if msg.tp != SysCtrlType.VALUE:
                return None
            if msg.cmd != cmd:
                return None
            return msg

        req = SysCtrlMessage()
        req.cmd = cmd
        req.src = src
        req.dst = dst
        req.tp = SysCtrlType.QUERY
        if data is None:
            data = b""
        req.data = data

        with self.can_bus.gct().map(flt) as listener:
            await self.can_bus.send(req)
            msg = await listener.next(timeout=0.5)
        assert isinstance(msg, SysCtrlMessage)
        return msg.data

    async def fetch_readings(
        self,
        src: int,
        group_idx: int,
        idx: Union[int, Iterable[int]],
        dst: int = BROADCAST_ADDR,
    ) -> List[MonitoringData]:
        """
        Fetch readings from the device.
        In case not all requested readings are received, this function times out.

        :param src: The source nodeid
        :param dst: The destination nodeid, default to the broadcast address
        :param group_idx: The group index to fetch
        :param idx: Either one reading index or an iterable thereof
        :return: A list of MonitoringData messages, sorted by reading index
        """
        unified_idx: Set[int] = set()
        if not is_iterable(idx):
            unified_idx = {idx}  # type: ignore
        else:
            unified_idx = set(idx)  # type: ignore

        req = MonitoringRequest()
        req.src = src
        req.dst = dst
        req.group_idx = group_idx
        for x in unified_idx:
            req.readings |= 1 << x

        def flt(msg: GctMessage) -> Optional[MonitoringData]:
            if not isinstance(msg, MonitoringData):
                return None
            if msg.src != dst and dst == BROADCAST_ADDR:
                return None
            if msg.group_idx == group_idx and msg.reading_idx in unified_idx:
                return msg
            return None

        with self.can_bus.gct().map(flt) as listener:
            await self.can_bus.send(req)
            return await asyncio.wait_for(
                self._receive_readings(listener, len(unified_idx)), timeout=0.1
            )

    async def _receive_readings(
        self, listener: Receiver, num_unique: int
    ) -> List[MonitoringData]:
        ret = {}
        while True:
            msg = await listener.next()
            assert isinstance(msg, MonitoringData)
            ret[msg.reading_idx] = msg
            if len(ret) == num_unique:
                break
        return list(sorted(list(ret.values()), key=lambda x: x.reading_idx))

    async def read_single(self, src: int, group_idx: int, idx: int) -> MonitoringData:
        return (await self.fetch_readings(src, group_idx, [idx]))[0]

    async def read_single_and_decode(
        self, src: int, group_idx: int, idx: int, format: str
    ) -> tuple:
        data = await self.read_single(src, group_idx, idx)
        return struct.unpack(format, data.data)

    async def ddp(
        self,
        src_addr: int,
        dst_addr: int,
        data: bytes,
        timeout: float,
    ) -> bytes | None:
        assert len(data) > 0
        msg = DdpMessage(version=self._version)
        data = bytearray(data)
        want_response = (data[0] & 0x80) != 0
        msg.data = bytes(data)
        msg.src = src_addr
        msg.dst = dst_addr

        flt = DdpFilter(src=dst_addr, dst=src_addr)
        with self.can_bus.gct().map(flt) as listener:
            await self.can_bus.send(msg)
            if not want_response:
                return None

            reply = await asyncio.wait_for(listener.next(), timeout)
            return reply

    async def ddp_request(
        self,
        cmd: int,
        src_addr: int,
        dst_addr: int,
        data: bytes,
        timeout: float = DEFAULT_DDP_TIMEOUT,
    ) -> bytes:
        assert cmd < 256, "Command must be a byte"
        data = bytes([cmd | 0x80]) + data
        reply = await self.ddp(src_addr, dst_addr, data, timeout)
        assert reply is not None
        return reply

    async def ddp_write(
        self,
        cmd: int,
        src_addr: int,
        dst_addr: int,
        data: bytes,
        timeout: float = DEFAULT_DDP_TIMEOUT,
    ) -> None:
        data = bytes([cmd & 0x7F]) + data
        await self.ddp(src_addr, dst_addr, data, timeout=timeout)


def is_iterable(arg: Any) -> bool:
    """
    Return True if the argument is an iterable. However strings are not considered an iterable
    since they usually used as primitives.
    """
    return isinstance(arg, Iterable) and not isinstance(arg, str)
